File: examples/euler1d.py
"""
Solves the 1d Euler equations using a first-order Godunov scheme

When run with Jax, this code performs at roughly 20 Mzps on a single-vapor CPU,
and about 2.6 Gzps on an A100. See the sister code euler1d.cpp for the C++
implementation. That code is typically 6x faster than Jax on the CPU and 4x
faster on the GPU.
"""
from time import perf_counter_ns

jax = True

if jax:
    from jax import jit, numpy as np

else:
    import numpy as np

    def jit(f):
        return f


gamma: float = 5.0 / 3.0


@jit
def cons_to_prim(u):
    rho = u[0]
    mom = u[1]
    nrg = u[2]
    p0 = rho
    p1 = mom / rho
    p2 = (nrg - 0.5 * mom**2 / rho) * (gamma - 1.0)
    return np.stack([p0, p1, p2], axis=0)


@jit
def prim_to_cons(p):
    rho = p[0]
    vel = p[1]
    pre = p[2]
    u0 = rho
    u1 = rho * vel
    u2 = 0.5 * rho * vel**2 + pre / (gamma - 1.0)
    return np.stack([u0, u1, u2], axis=0)


@jit
def prim_and_cons_to_flux(p, u):
    vel = p[1]
    pre = p[2]
    nrg = u[2]
    f0 = vel * u[0]
    f1 = vel * u[1] + pre
    f2 = vel * (nrg + pre)
    return np.stack([f0, f1, f2], axis=0)


@jit
def sound_speed_squared(p):
    rho = p[0]
    pre = p[2]
    return gamma * pre / rho


@jit
def riemann_hlle(ul, ur):
    pl = cons_to_prim(ul)
    pr = cons_to_prim(ur)
    fl = prim_and_cons_to_flux(pl, ul)
    fr = prim_and_cons_to_flux(pr, ur)
    csl = np.sqrt(sound_speed_squared(pl))
    csr = np.sqrt(sound_speed_squared(pr))
    al0 = pl[1] - csl
    al1 = pl[1] + csl
    ar0 = pr[1] - csr
    ar1 = pr[1] + csr
    am = np.minimum(al0, ar0).clip(None, 0.0)
    ap = np.maximum(al1, ar1).clip(0.0, None)
    return (fl * ap - fr * am - (ul - ur) * ap * am) / (ap - am)


@jit
def initial_primitive(x):
    l = np.array([1.0, 0.0, 1.000])
    r = np.array([0.1, 0.0, 0.125])
    return ((x < 0.5)[:, None] * l + (x > 0.5)[:, None] * r).T


def main():
    N = 100000

    xv = np.linspace(0.0, 1.0, N + 1)
    xc = 0.5 * (xv[1:] + xv[:-1])
    dx = (xv[-1] - xv[0]) / N
    dt = dx * 0.3

    u = prim_to_cons(initial_primitive(xc))
    n = 0
    t = 0.0
    t_final = 0.01
    plot = True
    fold = 20

    while t < t_final:
        t0 = perf_counter_ns()
        for i in range(fold):
            fhat = riemann_hlle(u[:, :-1], u[:, 1:])
            du = np.diff(fhat, axis=1) * (-dt / dx)

            if jax:
                u = u.at[:, 1:-1].add(du)
            else:
                u[:, 1:-1] += du

            t += dt
            n += 1
        t1 = perf_counter_ns()
        print(f"[{n:04d}] t={t:.3f} Mzps={1e3 * N * fold / (t1 - t0):.2f}")

    if plot:
        from matplotlib import pyplot as plt

        p = cons_to_prim(u)
        plt.plot(xc, p[0, :])
        plt.show()


main()

File: examples/euler1d.cpp
/**
================================================================================
Copyright 2023 - 2024, Jonathan Zrake

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
================================================================================
*/




#include <chrono>
#include <cstdio>
#include <cmath>
#include "vapor/array.hpp"
#include "vapor/executor.hpp"

using namespace vapor;
#define gamma (5.0 / 3.0)
#define min2(a, b) ((a) < (b) ? (a) : (b))
#define max2(a, b) ((a) > (b) ? (a) : (b))
#define min3(a, b, c) min2(a, min2(b, c))
#define max3(a, b, c) max2(a, max2(b, c))

#ifdef __NVCC__
#define CACHE_FLUX false
#else
#define CACHE_FLUX true
#endif




int main()
{
    auto cons_to_prim = [] HD (dvec_t<3> u)
    {
        auto rho = u[0];
        auto mom = u[1];
        auto nrg = u[2];
        auto p0 = rho;
        auto p1 = mom / rho;
        auto p2 = (nrg - 0.5 * mom * mom / rho) * (gamma - 1.0);
        return vec(p0, p1, p2);
    };

    auto prim_to_cons = [] HD (dvec_t<3> p)
    {
        auto rho = p[0];
        auto vel = p[1];
        auto pre = p[2];
        auto u0 = rho;
        auto u1 = rho * vel;
        auto u2 = 0.5 * rho * vel * vel + pre / (gamma - 1.0);
        return vec(u0, u1, u2);
    };

    auto prim_and_cons_to_flux = [] HD (dvec_t<3> p, dvec_t<3> u)
    {
        auto vel = p[1];
        auto pre = p[2];
        auto nrg = u[2];
        auto f0 = vel * u[0];
        auto f1 = vel * u[1] + pre;
        auto f2 = vel * (nrg + pre);
        return vec(f0, f1, f2);
    };

    auto sound_speed_squared = [] HD (dvec_t<3> p)
    {
        auto rho = p[0];
        auto pre = p[2];
        return gamma * pre / rho;
    };

    auto riemann_hlle = [=] HD (dvec_t<3> ul, dvec_t<3> ur)
    {
        auto pl = cons_to_prim(ul);
        auto pr = cons_to_prim(ur);
        auto fl = prim_and_cons_to_flux(pl, ul);
        auto fr = prim_and_cons_to_flux(pr, ur);
        auto csl = sqrt(sound_speed_squared(pl));
        auto csr = sqrt(sound_speed_squared(pr));
        auto alm = pl[1] - csl;
        auto alp = pl[1] + csl;
        auto arm = pr[1] - csr;
        auto arp = pr[1] + csr;
        auto am = min3(alm, arm, 0.0);
        auto ap = max3(alp, arp, 0.0);
        return (fl * ap - fr * am - (ul - ur) * ap * am) / (ap - am);
    };

    auto initial_primitive = [] HD (double x)
    {
        if (x < 0.5)
            return vec(1.0, 0.0, 1.0);
        else
            return vec(0.1, 0.0, 0.125);
    };
    auto exec = default_executor_t();
    auto pool = pool_allocator_t();

    auto t_final = 0.1;
    auto N = 20000;
    auto dx = 1.0 / N;
    auto iv = range(N + 1);
    auto ic = range(N);
    auto xc = (ic + 0.5) * dx;
    auto dt = dx * 0.3;
    auto u = xc.map(initial_primitive).map(prim_to_cons).cache(exec, pool);

    auto interior_faces = index_space(uvec(1), uvec(N - 1));
    auto interior_cells = index_space(uvec(1), uvec(N - 2));
    auto t = 0.0;
    auto n = 0;
    auto fold = 50;

    while (t < t_final)
    {
        auto t1 = std::chrono::high_resolution_clock::now();

        for (int m = 0; m < fold; ++m)
        {
            auto fhat = iv[interior_faces].map([u, riemann_hlle] HD (uint i)
            {
                auto ul = u[i - 1];
                auto ur = u[i];
                return riemann_hlle(ul, ur);
            }).cache(exec, pool);

            auto du = ic[interior_cells].map([fhat, dt, dx] HD (uint i)
            {
                auto fm = fhat[i];
                auto fp = fhat[i + 1];
                return (fp - fm) * (-dt / dx);
            });

            u = (u.at(interior_cells) + du).cache(exec, pool);

            t += dt;
            n += 1;
        }

        auto t2 = std::chrono::high_resolution_clock::now();
        auto delta = std::chrono::duration_cast<std::chrono::duration<double>>(t2 - t1);
        auto Mzps = N * fold * 1e-6 / delta.count();
        printf("[%04d] t=%.3lf Mzps=%.3lf\n", n, t, Mzps);
    }

    // auto p = u.map(cons_to_prim);
    // for (int i = 0; i < N; ++i)
    // {
    //     printf("%+.4f %+.4f %+.4f %+.4f\n", xc[i], p[i][0], p[i][1], p[i][2]);
    // }
    return 0;
}
